### ISSUE TITLE
Unable to use way.id inside way_function

### ISSUE THREAD

**Issue by p4bloch** [Project-OSRM/osrm-backend#1776]
I need to check every `way_id` against a postgresql database in order to add weight to it and make the router avoid it.

The problem is, inside way_function, I am not able to use `way.id`. This is the error I get when I try to print it:
~~~
terminate called after throwing an instance of 'osrm::exception'
  what():  ERROR occured in profile script:
No matching overload found, candidates:
void print(std::string)
~~~

And when trying to concatenate it like this `local sql_query = "SELECT rating FROM bk_rating WHERE way_id = " .. way.id .. " "`, I get the following error
~~~
terminate called after throwing an instance of 'osrm::exception'
  what():  ERROR occured in profile script:
profile.lua:170: attempt to concatenate field 'id' (a function value)
~~~ 

I've been trying to find answers to this error in the Lua community but none of the solutions seems to apply for OSRM. Is the `way.id` actually available in the `way` that is passed as a parameter to `way_function`? Am I calling it the right way? I took the query from [this postgis.lua example](https://github.com/springmeyer/Project-OSRM/blob/master/profiles/examples/postgis.lua)

Best,
Pablo

**Comment by danpat** [comment]
The way id is not exposed to Lua.  In general, we've found that ways aren't a great unit-of-measure for roads, some are really short, some are really long.

We're about to push some code into `develop` for the next 4.9.0 release that exposes some tools for fine-grained updating of edge weights.

Can you describe what you're trying to do?  It might be that the upcoming changes might be a better way to do it.

**Comment by p4bloch** [comment]
The idea is to use a custom value provided by an **external source** (not an OSM tag) to let OSRM know which roads it should avoid for cyclist (big avenues, bridges, etc).

I'm not a geo expert so the whole external raster data from the wiki was a bit too much for me, and using a postgresql database to link streets to our custom penalization value seemed like a good solution.

What do you recommend to achieve this?

Thank you

**Comment by emiltin** [comment]
have the access to the osm id been removed from lua? we depend on it for prioritizing ways depending on surrounding green areas (via postgis)

**Comment by danpat** [comment]
Sorry, my mistake, @lbud pointed out you can get it with `way:id()`, it's a function, not a property.

@p4bloch There are a few traps in the approach, not the least of which is that way ids change over time.  If you stick with a static copy of OSM, then you'll be fine, but if you want to pull in updates, you'll need to be prepared for way ids of objects to change, ways to get split/combined and generally do unexpected things.  If you're just looking for a limited region, and you want to bring in updates from OSM, you might be better off doing fuzzy tag comparisons (looking for bridges, matching on "*Brooklyn*" or something).

**Comment by emiltin** [comment]
what we do is update the postgis database everytime we pull in new data to osrm. so the same data is is in osrm and postgis and all ids match.

**Comment by p4bloch** [comment]
@danpat thanks for the insight on how `way_id` works within OSM. I naively expected that ids would behave like ids everywhere else and remain the same over time ¯\\_(ツ)_/¯. Using some string manipulation on the information OSM provides, although a bit restrictive, seems to be the way to go, right?

@emiltin Interesting approach. Do you simply replicate the data in your local postgis or do you keep custom columns with values different than OSM? I imagine you would need to do some mapping between the old values and the new ones in every update, right? If you could describe a little bit more about this process it'd be super helpful

I'd rephrase my question to be a bit more generic, given I was going the wrong way.. **If you needed to tell OSRM to avoid certain street or avenue between X and Y streets, how would you go about it?**

**Comment by p4bloch** [comment]
@danpat our point of view is rather subjective on how to tackle this issue, and may not be suitable for the whole community, that's one reason we don't want to submit this to OSM just yet. We are going to give our users what **we think** is the better route.

Also, we are going to try an approach which seems easier to implement in scale for us, that is using a custom "rating" for each relevant street. This approach is [explicitly discouraged](http://wiki.openstreetmap.org/wiki/Talk:Proposed_features/Cycleworth) by OSM's community, but we believe is the best fit for our needs and our capacity to port our user's insight into the router.

Those are the two reasons we want to stay away from contributing to OSM until we know for sure that our insight is really accurate and valuable for everyone else.

It'd be great to read more about how we can use coordinates to accurately block a street on OSRM, as it seems a very reliable solution. Any links/docs you could point me to?

Thanks!

**Comment by emiltin** [comment]
map matching is not trivial. have a look at openLR.

we use minutely updates to keep an pbf file up to date locally. we them
1) update a postgis db and run spatial analysis, computing things like 'green score' for all ways.
2) run osrm processing, using osm ids to reference/query data in postgis from lua.

this works because the exact same ways are in postgis and osrm and everything is based on data already in osm, eg. areas tagged as parks, lakes, etc. we can do things like prioritize proximity to forests, avoid paths close to big roads, etc, which could be seen as subjective but is based on objective osm data.

**Comment by TheMarex** [comment]
Closing here as the actual issue has been resolved.

**Comment by Spfeil2** [comment]
I have a similar problem. I would like to test postgis.lua functions but get the following error:

`terminate called after throwing an instance of 'sol::error'
  what():  lua: error: ../osrm-backend/profiles/postgis.lua:72: attempt to concatenate a function value (field 'id')
Aborted (core dumped)
`
I use the following code:
`function way_function (way, result)
  -- only route on ways with highway=*
  --local highway = way.tags:Find("highway") -- with this i would get an error
  local highway = way:get_value_by_key('highway')
  if (not highway or highway=='') then
    return 0
  end`

`local sql_query = " " ..
    "SELECT SUM(SQRT(area.area)) AS val " ..
    "FROM import.osm_roads way " ..
    "LEFT JOIN import.osm_landusages area ON ST_DWithin(way.geometry, area.geometry, 100) " ..
    "WHERE area.type IN ('industrial') AND way.osm_id=" .. way.id .. " " ..
    "GROUP BY way.id"`
from [local sql_query = " " ..
    "SELECT SUM(SQRT(area.area)) AS val " ..
    "FROM import.osm_roads way " ..
    "LEFT JOIN import.osm_landusages area ON ST_DWithin(way.geometry, area.geometry, 100) " ..
    "WHERE area.type IN ('industrial') AND way.osm_id=" .. way.id .. " " ..
    "GROUP BY way.id"](url)

when I change `way.id` to `way:id()`, I get the following error:
`Segmentation fault (core dumped)`

Best,
Simon
